### PR TITLE
feat: Proxy contract registry

### PR DESCRIPTION
**Pull request by hacheigriega**
## Explanation of Changes

This PR creates a transaction endpoint for submitting a proposal for (1) instantiating a contract using a predictable address and then (2) storing the address in the Proxy Contract registry. Technically, using a predictable address instantiation is unnecessary here, so we can choose to go with standard instantiation instead.

In a local setting (set up using `./scripts/local_setup.sh`), the new feature can be tested as follows:
```
BIN=./build/seda-chaind
ADDR=$($BIN keys show satoshi --keyring-backend test -a)
CONTRACT_PATH=./wasm/sample.wasm
INITIAL_STATE="{}"

# Deploy (store) contract
CONTRACT_DEPLOY_OUTPUT=$($BIN tx wasm store $CONTRACT_PATH --from $ADDR --keyring-backend test --gas 2000000 -y --output json)
CONTRACT_DEPLOY_TX_HASH=$(echo $CONTRACT_DEPLOY_OUTPUT | jq -r .txhash)
sleep 5;


# Instantiate deployed contract
CONTRACT_DEPLOY_TX_OUTPUT=$($BIN query tx $CONTRACT_DEPLOY_TX_HASH --output json)
CONTRACT_DEPLOY_CODE_ID=$(echo $CONTRACT_DEPLOY_TX_OUTPUT | jq -r '.events[] | select(.type | contains("store_code")).attributes[] | select(.key | contains("code_id")).value')
CONTRACT_DEPLOY_CODE_HASH=$(echo $CONTRACT_DEPLOY_TX_OUTPUT | jq -r '.events[] | select(.type | contains("store_code")).attributes[] | select(.key | contains("code_checksum")).value')

# (how to get predictable address in advance)
GOV_ADDR=seda10d07y265gmmuvt4z0w9aw880jnsr700jvvla4j # since gov module will send the instantiation tx
$BIN query wasm build-address $CONTRACT_DEPLOY_CODE_HASH $GOV_ADDR $(echo -n "testing" | xxd -ps) $INITIAL_STATE

$BIN tx wasm-storage submit-proposal instantiate-and-register-proxy-contract $CONTRACT_DEPLOY_CODE_ID $INITIAL_STATE $(echo -n "testing" | xxd -ps) \
	--label "local0.1.0" --admin $ADDR  \
	--fix-msg \
	--title nice_title --summary nice_summary --deposit 10000000aseda \
	--gas auto --gas-adjustment 1.2 --from $ADDR --keyring-backend test

sleep 10;

# Gov process
$BIN tx gov vote 1 yes --from $ADDR --keyring-backend test --gas auto --gas-adjustment 1.2 -y
```

Once the gov proposal is passed and executed, you can find that the registry is updated with the instantiated contract's address:
```
$BIN query wasm-storage proxy-contract-registry
```

 
## Testing

Tested in a local environment manually.

## Related PRs and Issues

Closes: Issue #47

### CODE REVIEW

**Review by Thomasvdam — COMMENTED** [review]
I'll be honest: I need to get more familiar with the project before I feel comfortable approving PRs. :D 
It doesn't always look clear to me, but I think that's more so my unfamiliarity with the SDK than any issues with the code.

**Review by JeancarloBarrios — CHANGES_REQUESTED** [review]
Hey, I think it broke the test. Can you run all the keeper tests? But its because don't have actions to run it for us at the PR level.

**Comment by hacheigriega** [comment]
@Thomasvdam Thanks for taking the time to go through the code nonetheless.
@JeancarloBarrios I quickly fixed the test with a nil keeper since I wasn't sure how to mock wasm `ContractOpsKeeper`..

File: x/wasm-storage/keeper/msg_server.go [hunk 1]
```diff
@@ -75,6 +77,47 @@ func (k msgServer) StoreOverlayWasm(goCtx context.Context, msg *types.MsgStoreOv
...
+			return nil, fmt.Errorf("invalid admin address: %s", err)
+		}
+	}
+
+	// policy := m.selectAuthorizationPolicy(ctx, msg.Sender)
+	// addrGenerator := wasmkeeper.PredicableAddressGenerator(senderAddr, msg.Salt, msg.Msg, msg.FixMsg)
+	// contractAddr, data, err := m.keeper.instantiate(ctx, msg.CodeID, senderAddr, adminAddr, msg.Msg, msg.Label, msg.Funds, addrGenerator, policy)
+	// if err != nil {
+	// return nil, err
+	// }
```
[the remark this answers is not in the document]
Comment (hacheigriega) [reply x/wasm-storage/keeper/msg_server.go]: nope - removed!

**Review by gluax — APPROVED** [review]
So if I'm understanding correctly, this is for being able to update the proxy address in gov? Not instantiating it before the network runs right? But I notice it's admin only(I think)? Don't we want anyone to be able to update the proxy for it to be permissionless? Or did I misunderstand and this for launching the initial proxy contract after we start the network?

CC @mariocao @FranklinWaller

**Comment by hacheigriega** [comment]
@gluax Yes, you understood correctly. It's not admin-only though (we don't even have the concept of admin on chain). Perhaps the authority check part confused you, but that is just to ensure that the transaction comes from the governance module. You got me wondering how we know governance module signature cannot be forged, though..

**Comment by JeancarloBarrios** [comment]
> @Thomasvdam Thanks for taking the time to go through the code nonetheless. @JeancarloBarrios I quickly fixed the test with a nil keeper since I wasn't sure how to mock wasm `ContractOpsKeeper`..

I will add this in a separate PR!!

**Review by JeancarloBarrios — APPROVED** [review]

**Review by mariocao — APPROVED** [review]
Thanks for adding the testing steps!

**Pull request merged**